scipy/optimize/_shuffled_complex_evolution.py

```diff
@@ -0,0 +1,1083 @@
+    * Warn only if lb > ub, simply set mask if lb == ub,
+      May 2023, Matthias Cuntz
+    * Rename sce to shuffled_complex_evolution and SCESolver to
+      ShuffledComplexEvolutionSolver, May 2023, Matthias Cuntz
+    * Exit if initial population failed twice, May 2023, Matthias Cuntz
+
+"""
+import warnings
+import numpy as np
+from scipy.optimize import OptimizeResult, minimize
+from scipy._lib._util import check_random_state
+from scipy.optimize._constraints import Bounds
+
+# ToDo:
+# - write tmp/population files (of Fortran code)
+
+__all__ = ['shuffled_complex_evolution']
+
+
+# extended scipy/_util.py version
+class _FunctionWrapper:
+    """
+    Wrap user function with arguments and keywords, allowing picklability
+
+    Parameters
+    ----------
+    func : callable
+        Function in the form ``func(x, *args, **kwargs)``, where ``x`` are
+        the parameters in the form of an iterable.
+        ``args`` and ``kwargs`` are passed to the function via the usual
+        unpacking operators.
+
+    """
+    def __init__(self, func, *args, **kwargs):
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self.nargs = len(args)
+        self.nkwargs = len(kwargs)
+        if self.nkwargs == 0:
+            if ( (len(self.args) == 2) and
+                 isinstance(self.args[-1], dict) and
+                 (len(self.args[-1]) == 0) ):
+                # if kwargs={} then **kwargs={} and hence counted as args
+                self.args = self.args[0]
+                self.nargs = len(args)
+
+    def __call__(self, x):
+        if (self.nargs > 0) and (self.nkwargs > 0):
+            return self.func(x, *self.args, **self.kwargs)
+        elif (self.nargs > 0) and (self.nkwargs == 0):
+            return self.func(x, *self.args)
+        elif (self.nargs == 0) and (self.nkwargs > 0):
+            return self.func(x, **self.kwargs)
+        else:
+            return self.func(x)
+
+
+# from distutils.util (Python 3.11.1)
+def _strtobool(val):
+    """
+    Convert a string representation of truth to true (1) or false (0).
+    True values are 'y', 'yes', 't', 'true', 'on', and '1'; false values
+    are 'n', 'no', 'f', 'false', 'off', and '0'.  Raises ValueError if
+    'val' is anything else.
+
+    """
+    val = val.lower()
+    if val in ('y', 'yes', 't', 'true', 'on', '1'):
+        return 1
+    elif val in ('n', 'no', 'f', 'false', 'off', '0'):
+        return 0
+    else:
+        raise ValueError("invalid truth value {!r}".format(val))
+
+
+def shuffled_complex_evolution(
+        func, x0, lb, ub=None,
+        mask=None,
+        args=(), kwargs={},
+        sampling='half-open',
+        maxn=1000, kstop=10, pcento=0.0001, peps=0.001,
+        ngs=2, npg=0, nps=0, nspl=0, mings=0,
+        seed=None, iniflg=True,
+        alpha=0.8, beta=0.45, maxit=False, printit=2,
+        polish=True,
+        restart=False, restartfile1='',
+        restartfile2=''):
```

Comment:
I think we should drop the `restart` functionality. As @rkern also mentioned, this is a generally useful feature for all optimizers, so it should not be available only for one. The API would have to be discussed further if this is generally desired among `scipy.optimize`.